galsim/config/gsobject.py

```diff
@@ -79,25 +61,49 @@ def BuildGSObject(config, key, base=None, gsparams={}, logger=None):
         param = config[key]
     except KeyError:
         return None, True
-
     if False:
         logger.debug('obj %d: param = %s',base['obj_num'],param)
 
-    # Check that the input config has a type to even begin with!
+    # Save these, so we can edit them based on parameters at this level in the tree to take 
+    # effect on all lower branches, and then we can reset it back to this at the end.
+    orig_index_key = base.get('index_key',None)
+    orig_rng = base.get('rng',None)
+
+    # Check what index key we want to use for this object.
+    # Note: this call will also set base['index_key'] and base['rng'] to the right values
+    index = galsim.config.value._get_index(param, base)
+
+    # Get the type to be parsed.
     if not 'type' in param:
         raise AttributeError("type attribute required in config.%s"%key)
     type_name = param['type']
 
-    # If we have previously saved an object and marked it as safe, then use it.
-    if 'current_val' in param and param['current_safe']:
-        if logger and logger.isEnabledFor(logging.DEBUG):
-            logger.debug('obj %d: current is safe: %s',base['obj_num'],str(param['current_val']))
-        return param['current_val'], True
+    # If we are repeating, then we get to use the current object for repeat times.
+    if 'repeat' in param:
+        repeat = galsim.config.ParseValue(param, 'repeat', base, int)[0]
+    else:
+        repeat = 1
 
-    # Ring is only allowed for top level gal (since it requires special handling in 
-    # multiprocessing, and that's the only place we look for it currently).
-    if type_name == 'Ring' and key != 'gal':
-        raise AttributeError("Ring type only allowed for top level gal")
+    # Check if we can use the current cached object
+    if ('current_val' in param and 
+            (param['current_safe'] or param['current_index']//repeat == index//repeat)):
+        # If logging, explain why we are using the current object.
+        if logger and logger.isEnabledFor(logging.DEBUG):
```

Comment:
I agree.  I made a LoggerWrapper that wraps the multiprocessing proxy object and does these checks before calling the proxy object's function.  So most places in the code can skip the `isEnabledFor` part of this line.  f62f5f47e1102a2180bfe902b4755c9b86371301